File: src/node/core/leaderboard.ts
import type { UsageEntry, UsageTotals } from "./types.js";
import type { BarMetric } from "../tui/formatter.js";
import { metricValue } from "../tui/formatter.js";
import { currentLabel } from "./fetcher.js";

// Pure leaderboard aggregation for the `lb` / `lbh` displays (intake §2).
// No I/O, no mutation of inputs (Constitution V) — the dispatch layer fetches
// and windows daily entries, this module ranks them.

export interface LeaderboardRow {
  rank: number;          // 1-based, after sorting
  user: string;          // user name, or "user/machine" under --by-machine
  machine?: string;      // present only under --by-machine
  totals: UsageTotals;   // summed across the source's tools and the window
  share: number;         // 0..1 of the grand total, in the display metric
  delta?: number;        // fractional change vs the previous window; undefined ⇒ "new"
}

// A resolved date window (ISO YYYY-MM-DD bounds) plus the label used in the
// Δ column header / heading. Bounds are optional only for an open-ended
// explicit --since/--until window.
export interface LeaderboardWindow {
  start?: string;
  end?: string;
  label: string;
}

const SHORT_MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
const DAY_MS = 86_400_000;

function isoLocal(d: Date): string {
  const mm = String(d.getMonth() + 1).padStart(2, "0");
  const dd = String(d.getDate()).padStart(2, "0");
  return `${d.getFullYear()}-${mm}-${dd}`;
}

// Day arithmetic on date-only ISO labels, parsed as UTC midnight (the same
// treatment weekLabel gives daily labels — timezone-independent).
function addDays(iso: string, days: number): string {
  const d = new Date(`${iso}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

function diffDays(a: string, b: string): number {
  return Math.round((new Date(`${b}T00:00:00Z`).getTime() - new Date(`${a}T00:00:00Z`).getTime()) / DAY_MS);
}

// The window the leaderboard ranks: the current period (today / this
// Sunday-anchored week / this month), or the explicit --since/--until range
// when one is given (it replaces the period window). The window label feeds
// the heading: the period's current label, or the range itself.
export function currentWindow(period: string, since: string | undefined, until: string | undefined, now: Date = new Date()): LeaderboardWindow {
  if (since !== undefined || until !== undefined) {
    const label = since !== undefined && until !== undefined
      ? `${since} → ${until}`
      : since !== undefined ? `${since} →` : `→ ${until}`;
    return { start: since, end: until, label };
  }
  const today = currentLabel("daily", now);
  if (period === "monthly") {
    return { start: `${today.slice(0, 7)}-01`, end: today, label: currentLabel("monthly", now) };
  }
  if (period === "weekly") {
    const sunday = currentLabel("weekly", now);
    return { start: sunday, end: today, label: sunday };
  }
  return { start: today, end: today, label: today };
}

// The immediately preceding window of the same kind: previous calendar day /
// previous Sunday-anchored week / previous calendar month — or, under an
// explicit window, the equal-length range ending the day before `since`
// (label "prev"). Returns undefined when no well-defined previous window
// exists (an open-ended explicit window), in which case every row is "new".
export function previousWindow(period: string, since: string | undefined, until: string | undefined, now: Date = new Date()): LeaderboardWindow | undefined {
  if (since !== undefined || until !== undefined) {
    if (since === undefined) return undefined;
    const end = until ?? currentLabel("daily", now);
    const length = diffDays(since, end) + 1;
    if (length < 1) return undefined;
    return { start: addDays(since, -length), end: addDays(since, -1), label: "prev" };
  }
  if (period === "monthly") {
    const first = new Date(now.getFullYear(), now.getMonth() - 1, 1);
    const last = new Date(now.getFullYear(), now.getMonth(), 0);
    return { start: isoLocal(first), end: isoLocal(last), label: SHORT_MONTHS[last.getMonth()] };
  }
  if (period === "weekly") {
    const sunday = currentLabel("weekly", now);
    const start = addDays(sunday, -7);
    return { start, end: addDays(sunday, -1), label: start };
  }
  const prev = new Date(now);
  prev.setDate(prev.getDate() - 1);
  const label = isoLocal(prev);
  return { start: label, end: label, label };
}

// Sum every key's entries into one UsageTotals (all six numeric fields).
// Pure: entries and the input map are never mutated.
export function sumByKey(byUser: Map<string, UsageEntry[]>): Map<string, UsageTotals> {
  const out = new Map<string, UsageTotals>();
  for (const [key, entries] of byUser) {
    const totals: UsageTotals = { totalCost: 0, inputTokens: 0, outputTokens: 0, cacheCreationTokens: 0, cacheReadTokens: 0, totalTokens: 0 };
    for (const e of entries) {
      totals.totalCost += e.totalCost;
      totals.inputTokens += e.inputTokens;
      totals.outputTokens += e.outputTokens;
      totals.cacheCreationTokens += e.cacheCreationTokens;
      totals.cacheReadTokens += e.cacheReadTokens;
      totals.totalTokens += e.totalTokens;
    }
    out.set(key, totals);
  }
  return out;
}

// Rank keys descending by the display metric. Keys with zero tokens AND zero
// cost in the window are dropped (mirrors the snapshot renderer's zero-row
// omission); ties break by key name ascending for deterministic output.
// `share` is the row's fraction of the grand total in the display metric
// (0 when the grand total is 0). `delta` is the fractional change vs the
// previous window; undefined (rendered "new") when the key had no previous
// window or a previous value of exactly 0 (no divide-by-zero).
export function buildLeaderboard(
  byUser: Map<string, UsageEntry[]>,
  prevByUser: Map<string, UsageEntry[]> | undefined,
  metric: BarMetric,
): LeaderboardRow[] {
  const summed = sumByKey(byUser);
  const prevSummed = prevByUser !== undefined ? sumByKey(prevByUser) : undefined;

  const kept = [...summed.entries()].filter(([, t]) => !(t.totalTokens === 0 && t.totalCost === 0));
  kept.sort((a, b) => metricValue(b[1], metric) - metricValue(a[1], metric) || (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));

  let grand = 0;
  for (const [, t] of kept) grand += metricValue(t, metric);

  return kept.map(([key, totals], i) => {
    const value = metricValue(totals, metric);
    const prevTotals = prevSummed?.get(key);
    const prevValue = prevTotals !== undefined ? metricValue(prevTotals, metric) : 0;
    const slash = key.indexOf("/");
    return {
      rank: i + 1,
      user: slash === -1 ? key : key.slice(0, slash),
      ...(slash === -1 ? {} : { machine: key.slice(slash + 1) }),
      totals,
      share: grand > 0 ? value / grand : 0,
      delta: prevValue !== 0 ? (value - prevValue) / prevValue : undefined,
    };
  });
}

File: src/go/internal/command/leaderboard.go
package command

import (
	"math"
	"sort"
	"strings"
	"time"

	"github.com/sahil87/tu/internal/config"
	"github.com/sahil87/tu/internal/fact"
	"github.com/sahil87/tu/internal/query"
	"github.com/sahil87/tu/internal/render/ansi"
	"github.com/sahil87/tu/internal/render/csv"
	renderjson "github.com/sahil87/tu/internal/render/json"
	"github.com/sahil87/tu/internal/render/markdown"
	"github.com/sahil87/tu/internal/view"
)

// leaderboardWindow is a resolved date window (ISO YYYY-MM-DD bounds, ""
// open) plus the label the heading / Δ header shows (the TS
// LeaderboardWindow).
type leaderboardWindow struct {
	start, end, label string
}

// shortMonths labels the monthly previous window (the TS SHORT_MONTHS).
var shortMonths = []string{"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"}

// addDays is day arithmetic on date-only ISO labels, parsed as UTC midnight
// (the TS addDays — timezone-independent, like WeekLabel).
func addDays(iso string, days int) string {
	d, err := time.Parse("2006-01-02", iso)
	if err != nil {
		return iso
	}
	return d.AddDate(0, 0, days).Format("2006-01-02")
}

// diffDays is the whole-day distance a → b on UTC-parsed labels (the TS
// diffDays: Math.round over the millisecond difference).
func diffDays(a, b string) int {
	da, errA := time.Parse("2006-01-02", a)
	db, errB := time.Parse("2006-01-02", b)
	if errA != nil || errB != nil {
		return 0
	}
	return int(math.Round(db.Sub(da).Hours() / 24))
}

// leaderboardWindows computes the current and previous ranking windows (the
// TS currentWindow/previousWindow, intake §3). An explicit --since/--until
// REPLACES the period window (heading label "S → U" / "S →" / "→ U"); its
// previous window is the equal-length range ending the day before S (label
// "prev", nil when only --until is given or the length is < 1). Period
// windows anchor on now in LOCAL time; the previous window is the previous
// calendar day / Sunday-anchored week / calendar month (labelled with the
// English 3-letter month of its last day).
func leaderboardWindows(period query.Period, since, until string, now time.Time) (cur leaderboardWindow, prev *leaderboardWindow) {
	if since != "" || until != "" {
		switch {
		case since != "" && until != "":
			cur = leaderboardWindow{since, until, since + " → " + until}
		case since != "":
			cur = leaderboardWindow{since, "", since + " →"}
		default:
			cur = leaderboardWindow{"", until, "→ " + until}
		}
		if since == "" {
			return cur, nil
		}
		end := until
		if end == "" {
			end = query.CurrentLabel(query.Daily, now)
		}
		length := diffDays(since, end) + 1
		if length < 1 {
			return cur, nil
		}
		return cur, &leaderboardWindow{addDays(since, -length), addDays(since, -1), "prev"}
	}

	today := query.CurrentLabel(query.Daily, now)
	switch period {
	case query.Monthly:
		cur = leaderboardWindow{today[:7] + "-01", today, query.CurrentLabel(query.Monthly, now)}
		y, m, _ := now.Date()
		first := time.Date(y, m-1, 1, 0, 0, 0, 0, now.Location())
		last := time.Date(y, m, 0, 0, 0, 0, 0, now.Location())
		return cur, &leaderboardWindow{
			first.Format("2006-01-02"), last.Format("2006-01-02"),
			shortMonths[int(last.Month())-1],
		}
	case query.Weekly:
		sunday := query.CurrentLabel(query.Weekly, now)
		cur = leaderboardWindow{sunday, today, sunday}
		start := addDays(sunday, -7)
		return cur, &leaderboardWindow{start, addDays(sunday, -1), start}
	default:
		cur = leaderboardWindow{today, today, today}
		yesterday := now.AddDate(0, 0, -1).Format("2006-01-02")
		return cur, &leaderboardWindow{yesterday, yesterday, yesterday}
	}
}

// rankLeaderboard ranks the keys (user, or user/machine under --by-machine)
// over the current window (the TS sumByKey + buildLeaderboard, intake §3):
//
//   - The per-day collapse (Collapse over dims + Date) sums each key's tools
//     and machines per day in RECORD INPUT ORDER (the TS mergeEntries fold),
//     then SortByDate orders the days so the window sum folds ascending by
//     label — the JSON cost/share/delta doubles are byte surfaces (DC-10).
//   - Keys with TotalTokens == 0 && TotalCost == 0 in the window are dropped
//     (a key with previous-window data only is not in cur and drops too).
//   - Sort: descending by metricValue in the display metric, ties by key
//     ascending byte order (the TS string compare on ASCII keys).
//   - grand folds metricValue over the RANKED rows from 0 (order is
//     byte-visible in share); Share = value/grand, 0 when grand is 0.
//   - Delta = (value − prevValue)/prevValue when the previous-window value is
//     nonzero, else nil (rendered "new", JSON null, CSV empty).
func rankLeaderboard(raw []fact.Record, byMachine bool, cur leaderboardWindow, prev *leaderboardWindow, metric view.Metric) []view.LeaderboardRow {
	dims := []query.Dim{query.User}
	if byMachine {
		dims = append(dims, query.Machine)
	}
	daily := query.SortByDate(query.Collapse(raw, append(dims, query.Date)...))

	type keyed struct {
		key    string
		totals fact.Totals
	}
	kept := make([]keyed, 0)
	for _, g := range query.GroupBy(query.Window(daily, cur.start, cur.end), dims...) {
		if g.TotalTokens == 0 && g.TotalCost == 0 {
			continue
		}
		kept = append(kept, keyed{leaderboardGroupKey(g.Key, byMachine), g.Totals})
	}
	sort.Slice(kept, func(a, b int) bool {
		va := leaderboardMetricValue(kept[a].totals, metric)
		vb := leaderboardMetricValue(kept[b].totals, metric)
		if va != vb {
			return va > vb
		}
		return kept[a].key < kept[b].key
	})

	prevTotals := make(map[string]fact.Totals)
	if prev != nil {
		for _, g := range query.GroupBy(query.Window(daily, prev.start, prev.end), dims...) {
			prevTotals[leaderboardGroupKey(g.Key, byMachine)] = g.Totals
		}
	}

	grand := 0.0
	for _, k := range kept {
		grand += leaderboardMetricValue(k.totals, metric)
	}
	rows := make([]view.LeaderboardRow, len(kept))
	for i, k := range kept {
		value := leaderboardMetricValue(k.totals, metric)
		rows[i] = view.LeaderboardRow{Rank: i + 1, Totals: k.totals}
		if byMachine {
			slash := strings.Index(k.key, "/")
			rows[i].User, rows[i].Machine = k.key[:slash], k.key[slash+1:]
		} else {
			rows[i].User = k.key
		}
		if grand > 0 {
			rows[i].Share = value / grand
		}
		prevValue := 0.0
		if pt, ok := prevTotals[k.key]; ok {
			prevValue = leaderboardMetricValue(pt, metric)
		}
		if prevValue != 0 {
			d := (value - prevValue) / prevValue
			rows[i].Delta = &d
		}
	}
	return rows
}

// leaderboardGroupKey is a group's ranking key: the user, or "user/machine"
// under --by-machine (the TS map key).
func leaderboardGroupKey(key fact.Record, byMachine bool) string {
	if byMachine {
		return key.User + "/" + key.Machine
	}
	return key.User
}

// leaderboardMetricValue is the Totals field the leaderboard ranks, shares
// and deltas in (the TS metricValue): cost, or tokens under -t.
func leaderboardMetricValue(t fact.Totals, m view.Metric) float64 {
	if m == view.Tokens {
		return float64(t.TotalTokens)
	}
	return t.TotalCost
}

// runLeaderboard is the lb path (intake §11): the explicit window replaces
// the period window; the ranked list is computed in full and --top slices
// what each format EMITS (shares, the Total row and the Result totals cover
// the full set). Repo-only: raw is gatherAllUsers' output; no fetch, no
// warnings.
func runLeaderboard(req Request, cfg config.Config, raw []fact.Record, notices []string, deps Deps) Result {
	cur, prev := leaderboardWindows(req.Period, req.Flags.Since, req.Flags.Until, deps.Now())
	metric := view.Cost
	if req.Flags.Metric == Tokens {
		metric = view.Tokens
	}
	rows := rankLeaderboard(raw, req.Flags.ByMachine, cur, prev, metric)
	deltaLabel := "prev"
	if prev != nil {
		deltaLabel = prev.label
	}

	sliced := rows
	if req.Flags.Top > 0 && req.Flags.Top < len(rows) {
		sliced = rows[:req.Flags.Top]
	}

	res := Result{Notices: notices, CostByItem: make(map[string]float64, len(rows))}
	for _, r := range rows {
		res.TotalCost += r.TotalCost
		res.TotalTokens += r.TotalTokens
		res.CostByItem[leaderboardRowKey(r)] = leaderboardMetricValue(r.Totals, metric)
	}

	switch req.Format {
	case JSON:
		res.Lines = renderjson.Leaderboard(sliced)
	case CSV:
		res.Lines = csv.Leaderboard(sliced, rows, req.Flags.ByMachine)
	case Markdown:
		res.Lines = markdown.Leaderboard(sliced, rows, req.Period, deltaLabel, req.Flags.ByMachine)
	default:
		pinned := req.Flags.User
		if pinned == "" {
			pinned = cfg.User
		}
		res.Lines = ansi.Table(view.Leaderboard(rows, view.LeaderboardOptions{
			Period:      req.Period,
			WindowLabel: cur.label,
			DeltaLabel:  deltaLabel,
			Metric:      metric,
			PinnedUser:  pinned,
			Top:         req.Flags.Top,
			Width:       deps.Width,
			LastSync:    deps.lastSync(),
			Prev:        deps.live().Prev,
		}), deps.Colors)
	}
	return res
}

// leaderboardRowKey is a row's CostByItem/watch-map key (the TS
// leaderboardPrevMap): the user, or user/machine under --by-machine.
func leaderboardRowKey(r view.LeaderboardRow) string {
	if r.Machine != "" {
		return r.User + "/" + r.Machine
	}
	return r.User
}

// runLeaderboardHistory is the lbh path (intake §11): one view.Series per
// Repo.Users() profile, in that order, each entry a period label — the value
// per (user, label) is the LEFT FOLD in record input order over the
// windowed, relabelled records (tool-major, walk order within a tool: the TS
// aggregateMachineMap aggregates each tool's unmerged machine-major entries
// and sumLeaderboardToolMaps adds the tools in registry order) — NOT the main
// history's collapse-then-roll-up. Entries sort ascending by label after
// grouping (the TS mergeEntries sort); a user with no records in the window
// yields an empty series (kept — KeepAllColumns). Repo-only: no fetch, no
// warnings.
func runLeaderboardHistory(req Request, cfg config.Config, raw []fact.Record, notices []string, capActive bool, deps Deps) Result {
	metric := view.Cost
	if req.Flags.Metric == Tokens {
		metric = view.Tokens
	}
	series := make([]view.Series, 0)
	for _, u := range deps.Repo.Users() {
		recs := query.ByUser(raw, u) // gather order preserved
		groups := query.GroupBy(query.Relabel(query.Window(recs, req.Flags.Since, req.Flags.Until), req.Period), query.Date)
		entries := make([]view.Entry, len(groups))
		for i, g := range groups {
			entries[i] = view.Entry{Label: g.Key.Date, Totals: g.Totals}
		}
		sort.SliceStable(entries, func(a, b int) bool { return entries[a].Label < entries[b].Label })
		series = append(series, view.Series{Name: u, Entries: entries})
	}
	if req.Flags.Top > 0 {
		series = foldColumns(series, req.Flags.Top, metric)
	}

	// Result stats as runHistory computes them (the TS buildCostMap over the
	// folded map): {user}:{label} and total:{label} in the display metric.
	res := Result{Notices: notices, CostByItem: make(map[string]float64)}
	for _, s := range series {
		for _, e := range s.Entries {
			res.TotalCost += e.TotalCost
			res.TotalTokens += e.TotalTokens
			v := leaderboardMetricValue(e.Totals, metric)
			res.CostByItem[s.Name+":"+e.Label] = v
			res.CostByItem["total:"+e.Label] += v
		}
	}

	base := "Leaderboard History"
	if metric == view.Tokens {
		base = "Leaderboard Token History"
	}
	switch req.Format {
	case JSON:
		res.Lines = renderjson.TotalHistory(series)
	case CSV:
		res.Lines = csv.TotalHistory(series)
	case Markdown:
		res.Lines = markdown.TotalHistory(series, req.Period, capActive, base)
	default:
		// lbh carries Prev (the delta map) but no MaxRows (the TS
		// lbhFormatOptions never carries it) and no compact form.
		res.Lines = ansi.Table(view.TotalHistory(series, view.HistoryOptions{
			Period:          req.Period,
			Now:             deps.Now(),
			Width:           deps.Width,
			CapActive:       capActive,
			Metric:          metric,
			Prev:            deps.live().Prev,
			Title:           "📊 " + base + " (" + view.PeriodLabel(req.Period, capActive) + ")",
			RankColumns:     true,
			HighlightLeader: true,
			KeepAllColumns:  true,
		}), deps.Colors)
	}
	return res
}

// foldColumns is the lbh --top fold (the TS foldLeaderboardColumns): the
// series total is the left fold of the metric values over its entries
// (ascending) from 0; a stable descending sort picks the top N names; the
// kept series stay in their ORIGINAL order; the folded series' entries
// concatenate in original order and re-group per label (left fold in that
// order, sorted by label) into one "others" series appended LAST. n ≥ len
// folds nothing; an "others" with no entries is not appended.
func foldColumns(series []view.Series, top int, metric view.Metric) []view.Series {
	if top >= len(series) {
		return series
	}
	totals := make([]float64, len(series))
	for i, s := range series {
		for _, e := range s.Entries {
			totals[i] += leaderboardMetricValue(e.Totals, metric)
		}
	}
	order := make([]int, len(series))
	for i := range order {
		order[i] = i
	}
	sort.SliceStable(order, func(a, b int) bool { return totals[order[a]] > totals[order[b]] })
	keep := make(map[int]bool, top)
	for _, i := range order[:top] {
		keep[i] = true
	}
	out := make([]view.Series, 0, top+1)
	var folded []fact.Record
	for i, s := range series {
		if keep[i] {
			out = append(out, s)
			continue
		}
		for _, e := range s.Entries {
			folded = append(folded, fact.Record{Date: e.Label, Totals: e.Totals})
		}
	}
	if len(folded) == 0 {
		return out
	}
	groups := query.GroupBy(folded, query.Date)
	entries := make([]view.Entry, len(groups))
	for i, g := range groups {
		entries[i] = view.Entry{Label: g.Key.Date, Totals: g.Totals}
	}
	sort.SliceStable(entries, func(a, b int) bool { return entries[a].Label < entries[b].Label })
	return append(out, view.Series{Name: "others", Entries: entries})
}
